contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,348 @@
+
+```
+// TODO: Find a better name for it?
+type BulkSelector struct {
+	SingleTypeBulkSelector []SingleTypeBulkSelector
+}
+
+// TODO: Find a better name for it?
+type SingleTypeBulkSelector struct {
+	// FIXME: This should be meta.k8s.io/GroupVersionKind instead (or something
+	// similar to it).
+	ResourceType string
+
+	// We would like to reuse the same ListOptions definition as we are using
+	// in regular APIs. For the purpose of bulk operations, you can just specify
+	// more than one.
+	//
+	// NOTE: In the initial implementation, we will require all Options (for a
+	// given resource type) to have exactly the same ResourceVersion - otherwise
+	// we will reject it at the validation level.
+	// This will simplify the initial implementation as we need to guarantee that
+	// watch events within a single resource types are delivered with increasing
+	// resource versions, which may be more tricky with different resource versions.
+	Options []ListOptions
+}
+```
+
+ With such API in place, all List/Watch requests coming from kubelets will be
+allowed to List/Watch only objects that are "attached" to pods bound to a given
+node. We will create a dedicated admission plugin that will be responsible for
+detecting whether a given kubelet (user in general) is allowed to list/watch
+objects requested by `BulkSelector ` (the exact mechanism for it is out of
+scope for this doc) and either rejecting the whole request or letting it go (we
+are not going to support partial rejections (e.g. for this resource you can
+proceed for that one no).
+
+ With the above semantics, whenever user wants to start watching another object
+(or set of objects) or drop one (or a set), he needs to break the watch and
+initiate a new one with different selectors. That isn't perfect, but we still
+can solve all usecases with it.
+ What is most important is that a bunch of them can still be solved efficiently.
+As an example, we can solve watching secrets in Kubelet as following:
+- create a store that will also support the following checks on add/update/delete
+operations: "if the resource version of object is older that we is currently
+cached, ignore it"
+- use this store underneath the reflector framework (though reflector will
+require some changes)
+- whenever a set of secrets to watch changes:
+  - break the current watch
+  - retrieve current state of secrets that are new to be watched and add it to
+    the local cache
+  - remove from local cache secrets that are no longer supposed to be watched
+  - send new watch request with the updated filter (with RV from the moment when
+    we broke the watch)
+
+With that approach, even if watch is lagging and we will get some previous
+version of secret (after retrieving it), it will be ignored.
+
+TODO: We should also allow for breaking the watch if the ACLs change (and kubelet
+is no longer allowed to watch all requested objects).
+
+
+As a v2 version of this API, we can consider adding an ability to use websockets
```

Comment:
Let me straw man an implementation:

- Websocket upgrades require GET, which eliminates bulk up front send
- We want to keep the model as similar as possible to the bulk POST - clients shouldn't need to know new types
- This is very similar to pub/sub protocols and should seek to be familiar to existing client libraries and patterns

The point of a websocket streaming protocol is to allow web clients to easily start, stop, or restart watches for multiple resources as needed.  For example, consider a web dashboard for kubernetes.

1. User logs in to the console and selects a namespace from a list
2. The console shows an overview for that namespace, selecting multiple resources.  The console, wishing to be reactive, wants to start a watch for the most important resources, including deployments, pods, and statefulsets.
3. The user navigates into a particular deployment.  The deployment overview page wishes to begin watching for replica set changes that specifically correspond to that deployment.
4. The user selects the back button on their browser to return to the overview - the console needs to stop watching for changes to the deployment.
5. In the background, it has continued to update its caches for deployments, pods, and statefulsets.  It presents the set of accumulated changes to the user immediately, subtly highlighting the changes that occurred since the user was away.
6. The user clicks into a page that shows information about a third party resource.  During the watch, the third party resource server crashes and is restarted.  The console sees that an error occurred on that watch, and knows to begin attempting to LIST the new watch.

This implies a protocol such as:

1. Perform websocket connection upgrade to the bulk watch endpoint via GET
2. The client must **start** new watches, receive watch events and correlate them to a particular watch, detect the failure of a watch, and be able to **stop** an existing watch
3. We could support bulk watch start, or only singular start - the latter is simpler for most clients
4. We must be able to identify which watch an event correlates to - if we assume each watch has a unique identifier (like an HTTP2 stream or the existing channel semantics for our current streaming protocols), then we can add that identifier to the existing `Event` schema
5. When we start a watch, we must be able to receive or specify a new watch identifier.  The simplest option is to not allow specification, and instead have the server send a simple monotonically increasing integer for each new watch.  Clients must track that, but clients must track a mapping anyway
6. Starting a watch could involve sending an API object serialized that is similar to a single start from bulk watch (the `SingleTypeBulkSelector` or my `ListSelector` but with a single Options)
7. Today, any watch close is usually accompanied by an error type.  We could formalize this and state that any error closes a watch (which makes clients compatible with existing watches), and potentially ensure the returned api.Status include a reason.
8. Stopping a watch requires some new signal, so we have to either switch to a command / options structure (`START {...}` and `STOP {...}`), or come up with a way to represent that in an object
  - Command option is more similar to existing pub/sub, but we could have a `MaxResourceVersion` selector on the selector options that functions similarly.
  - Needs more thought

If all watches are closed, we could close the watch connection.